Spherical harmonic transform

tluebeck opened the pull request:
### Which issue(s) are closed by this pull request?

Closes #133

### Changes proposed in this pull request:

- add spherical harmonics transform and its inverse
- depends on #101 
- we need to decide where we implement it, probably shift to `transforms`?

tluebeck committed:
add sht

tluebeck committed:
rearrange sht

tluebeck committed:
Merge branch 'dev_spherical_harmonic_signal_class' into spherical_harmonics_transform

tluebeck committed:
update SHT

tluebeck committed:
Merge branch 'develop' into spherical_harmonics_transform

tluebeck committed:
update according to new SHSignal

tluebeck committed:
update sht

tluebeck committed:
edit doc strings

tluebeck committed:
edit docstring

tluebeck committed:
Merge branch 'develop' into spherical_harmonics_transform

tluebeck committed:
update tests

tluebeck committed:
Merge branch 'develop' into spherical_harmonics_transform

tluebeck committed:
update tests

tluebeck committed:
test back and forth for different normalizations, basis types, and condon shortley conventions

mberz commented on spharpy/sht.py:
```diff
@@ -0,0 +1,124 @@
+from . import SphericalHarmonics
+
+
+def sht(signal, coordinates, n_max, basis_type="real", axis=-2,
+        channel_convention='acn', normalization='n3d', condon_shortley='auto',
+        inverse_method='pseudo_inverse'):
```
```suggestion
def sht(signal, spherical_harmonics):
```
How about passing a SphericalHarmonic object here. This simplifies things a lot.

tluebeck replied:
I think it is easier for the user to only have one method which creates the SH class and performs the transform. Moreover, with the default parameters the user only needs to pass the signal, coordinates and n_max. But I thinks its better to discuss in the weekly.

mberz replied:
Yes, let's discuss during the weekly.

mberz reviewed (changes requested):
I think we should review the concept again. Specifying all properties again does not use the power of the SH class at all from a users perspective.
Passing the SH class instead of creating it within the function would make things much easier in my opinion.

tluebeck committed:
bugfix auto in inverse method in isft

tluebeck committed:
bugfix reshaping

f-brinkmann commented:
- We decided that the functions should get an SH class object.
- @tluebeck will check if it is easy to have SHTimeData and SHFrequencyData by changing inheritance.

tluebeck committed:
add SHAudio SHTimeData and SHFrequencyData

tluebeck committed:
add tests

tluebeck committed:
add doc and clean up

tluebeck committed:
fix typo

f-brinkmann commented on spharpy/sht.py:
```diff
@@ -0,0 +1,138 @@
+            Canada, May 2004, p. 45-48, doi: 10.1109/ICASSP.2004.1326759.
+    """
+
+    if not signal.time.shape[axis] == coordinates.csize:
+        if coordinates.csize not in signal.time.shape:
+            raise ValueError("Signal shape does not match "
+                             "number of coordinates.")
+        else:
+            axis = signal.time.shape.index(coordinates.csize)
+            warnings.warn("Compute spherical harmonics transform along "
+                          f"axis = {axis}.", UserWarning)
```
I think we should only raise an error here and not change the user input. But I like the idea of detecting the axis. How about having `axis='auto'` as a default and use

`axis = np.where(np.array(signal.cshape) == Y_inv.shape[1])[0]` 

to detect the axis. If `len(axis) == 1` we proceed if `len(axis) == 0` we raise an error and if `len(axis) > 1` we raise a different errror.

tluebeck committed:
restructure sht

tluebeck committed:
fix typo

tluebeck committed:
update according to review

add docstrings
move n_max and dimensions check to base class

tluebeck committed:
bugfix auto axis

tluebeck committed:
fix typo

tluebeck committed:
update according to review

tluebeck committed:
Merge branch 'develop' into dev/spherical_harmonics_audio

tluebeck committed:
adapt to new normalization names

tluebeck committed:
Merge branch 'develop' into dev/spherical_harmonics_audio

tluebeck committed:
add SphericalharmonicDefinition to SHAudio Classes

tluebeck committed:
ruff checks

tluebeck committed:
update according to review

tluebeck committed:
Merge branch 'develop' into dev/spherical_harmonics_audio

tluebeck committed:
update according ot review

tluebeck committed:
rename audio to sh_audio

tluebeck committed:
revert renaming of sh_audio

tluebeck committed:
Merge branch 'develop' into dev/spherical_harmonics_audio

tluebeck committed:
Incorporate new SHBase class

tluebeck committed:
override time and freq setter in SHaudio

tluebeck committed:
Change normalization and channel_order behaiviour

Always store fully normalized data in ACN order and renormalize accordinglz in setter/getter

tluebeck committed:
udpate according to review

tluebeck committed:
add data init methods

manually integrate branch #229

tluebeck committed:
clean up tests

tluebeck committed:
update tests

allow to pass sh data with dimensions < 3

tluebeck committed:
fix ruff errors

tluebeck committed:
clean up

tluebeck committed:
update according to review

tluebeck committed:
update according to review

overwrite basis_type setter and freq_raw

tluebeck committed:
update tests

tluebeck committed:
allow Signal, Frequency and TimeData

tluebeck committed:
Merge branch 'dev/spherical_harmonics_audio' into spherical_harmonics_transform

tluebeck committed:
merge sh audio classes

tluebeck committed:
adapt to new SH classes structure

tluebeck committed:
add input checks

tluebeck committed:
update tests

tluebeck committed:
update tests

tluebeck committed:
fix back and forward SHT tests

tluebeck committed:
clean up

tluebeck committed:
fix typos

tluebeck committed:
implement isht in/out tests

tluebeck committed:
Merge branch 'develop' into spherical_harmonics_transform

tluebeck committed:
refactor sht and isht

tluebeck committed:
bugfix wrong axis

tluebeck committed:
Merge branch 'develop' into spherical_harmonics_transform

tluebeck committed:
add dimension test and edit docstring

tluebeck committed:
merge develop

f-brinkmann reviewed (commented):
Thanks for the effort. Functionality is great to have and looks good to me. I only marked two points for reducing code for now.

tluebeck committed:
Merge branch 'develop' into spherical_harmonics_transform

tluebeck committed:
adapt to new class method

tluebeck committed:
replace tensordot with internal matrix_multiplication

tluebeck committed:
Merge branch 'develop' into spherical_harmonics_transform

tluebeck committed:
fix typo

tluebeck committed:
fix ruff errors

tluebeck committed:
fix ruff errors

tluebeck committed:
fix ruff errors

tluebeck committed:
fix ruff errors

f-brinkmann reviewed (commented):
Code and tests look good to me.

The new functions don't show up in the docs currently and I think it might be good to provide a mathematical definition of the SHT and ISHT somewhere. I see three options and list them in order of my preference. In any case, this could be done after merging the current state for east of review.

1. Add a spharpy.sht module, that contains sht and isht and a top-level module docstring.
2. Add a section 'Spherical harmonic transform' to the 'Definitions' sections of the spharpy docs and reference that section in sht and isht
3. Extend the documentation of sht and isht.

mberz commented:
To me adding it to the module docstring is probably the most intuitive version. 
I would include it in this PR, rather than a separate one.

tluebeck committed:
add sht to docs

tluebeck committed:
format docstring

tluebeck committed:
refine documentation

tluebeck commented:
I added the documentation to the top of the module and added it to the docs.

mberz commented:
waiting for the outcome of the discussion in #326
